File: apps/resonance/ResonanceCore/Sources/ResonanceCore/ManualContinueTrace.swift
import Foundation

/// Horodatage et delta inter-étapes pour le flux « J'ai ajouté le morceau, continuer ».
/// Écrit sur stderr : `manual-continue-trace: [HH:mm:ss.SSS] (+Nms) …`
public enum ManualContinueTrace {
    private static let lock = NSLock()
    private static var sessionID = ""
    private static var lastTimestamp: Date?

    private static let timeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .current
        formatter.dateFormat = "HH:mm:ss.SSS"
        return formatter
    }()

    public static func beginSession(importSessionID: String) {
        lock.lock()
        defer { lock.unlock() }
        sessionID = importSessionID
        lastTimestamp = nil
        emit("BEGIN session import_session_id=\(importSessionID)")
    }

    public static func log(_ message: String) {
        lock.lock()
        defer { lock.unlock() }
        emit(message)
    }

    private static func emit(_ message: String) {
        let now = Date()
        let deltaLabel: String
        if let lastTimestamp {
            let deltaMs = max(0, Int(now.timeIntervalSince(lastTimestamp) * 1000))
            deltaLabel = " (+\(deltaMs)ms)"
        } else {
            deltaLabel = ""
        }
        lastTimestamp = now
        let timestamp = timeFormatter.string(from: now)
        let sessionSuffix = sessionID.isEmpty ? "" : " [session=\(sessionID)]"
        let line = "manual-continue-trace: [\(timestamp)]\(deltaLabel)\(sessionSuffix) \(message)\n"
        if let data = line.data(using: .utf8) {
            FileHandle.standardError.write(data)
        }
    }
}

File: playlist_builder/infrastructure/manual_continue_trace.py
from __future__ import annotations

import sys
import time
from datetime import datetime
from threading import Lock

_TRACE_LOCK = Lock()
_TRACE_LAST_MONO: float | None = None
_TRACE_SESSION_ID = ""


def begin_session(import_session_id: str) -> None:
    global _TRACE_LAST_MONO, _TRACE_SESSION_ID
    with _TRACE_LOCK:
        _TRACE_SESSION_ID = import_session_id.strip()
        _TRACE_LAST_MONO = None
    log(f"BEGIN session import_session_id={import_session_id.strip()}")


def log(message: str) -> None:
    global _TRACE_LAST_MONO
    now_wall = datetime.now()
    now_mono = time.monotonic()
    with _TRACE_LOCK:
        if _TRACE_LAST_MONO is None:
            delta_label = ""
        else:
            delta_ms = max(0, int((now_mono - _TRACE_LAST_MONO) * 1000))
            delta_label = f" (+{delta_ms}ms)"
        _TRACE_LAST_MONO = now_mono
        session_suffix = f" [session={_TRACE_SESSION_ID}]" if _TRACE_SESSION_ID else ""
        timestamp = now_wall.strftime("%H:%M:%S.") + f"{now_wall.microsecond // 1000:03d}"
        line = f"manual-continue-trace: [{timestamp}]{delta_label}{session_suffix} {message}"
        print(line, file=sys.stderr, flush=True)
